File: packages/python/elizaos/features/basic_capabilities/providers/recent_messages.py
from __future__ import annotations

from typing import TYPE_CHECKING

from elizaos.generated.spec_helpers import require_provider_spec
from elizaos.types import Provider, ProviderResult

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime, Memory, State

# Get text content from centralized specs
_spec = require_provider_spec("RECENT_MESSAGES")


async def get_recent_messages_context(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> ProviderResult:
    room_id = message.room_id
    if not room_id:
        return ProviderResult(
            text="", values={"messageCount": 0, "hasHistory": False}, data={"messages": []}
        )

    sections: list[str] = []
    message_list: list[dict[str, str | int]] = []

    # Use lastCompactionAt as a lower bound so we skip compacted history
    start: int | None = None
    room = await runtime.get_room(room_id)
    if room and getattr(room, "metadata", None):
        start = room.metadata.get("lastCompactionAt")

    recent_messages = await runtime.get_memories(
        room_id=room_id,
        limit=20,
        order_by="created_at",
        order_direction="desc",
        **({"start": start} if start is not None else {}),  # type: ignore[arg-type]
    )

    recent_messages = list(reversed(recent_messages))

    for msg in recent_messages:
        if not msg.content or not msg.content.text:
            continue

        sender_name = "Unknown"
        if msg.entity_id:
            entity = await runtime.get_entity(msg.entity_id)
            if entity and entity.name:
                sender_name = entity.name
            elif str(msg.entity_id) == str(runtime.agent_id):
                sender_name = runtime.character.name

        message_text = msg.content.text
        if len(message_text) > 300:
            message_text = message_text[:300] + "..."

        msg_dict = {
            "id": str(msg.id) if msg.id else "",
            "sender": sender_name,
            "text": message_text,
            "timestamp": msg.created_at or 0,
        }
        message_list.append(msg_dict)
        sections.append(f"**{sender_name}**: {message_text}")

    context_text = "# Recent Messages\n" + "\n".join(sections) if sections else ""

    return ProviderResult(
        text=context_text,
        values={
            "messageCount": len(message_list),
            "hasHistory": len(message_list) > 0,
            "roomId": str(room_id),
        },
        data={
            "messages": message_list,
            "roomId": str(room_id),
        },
    )


recent_messages_provider = Provider(
    name=_spec["name"],
    description=_spec["description"],
    get=get_recent_messages_context,
    position=_spec.get("position"),
    dynamic=_spec.get("dynamic", True),
)

File: packages/rust/src/features/basic_capabilities/providers/recent_messages.rs
//! RECENT_MESSAGES provider implementation.

use async_trait::async_trait;
use once_cell::sync::Lazy;

use crate::error::PluginResult;
use crate::generated::spec_helpers::require_provider_spec;
use crate::runtime::IAgentRuntime;
use crate::types::{Memory, ProviderResult, State};

use super::Provider;

// Get text content from centralized specs
static SPEC: Lazy<&'static crate::generated::spec_helpers::ProviderDoc> =
    Lazy::new(|| require_provider_spec("RECENT_MESSAGES"));

/// Provider for recent message history.
pub struct RecentMessagesProvider;

#[async_trait]
impl Provider for RecentMessagesProvider {
    fn name(&self) -> &'static str {
        &SPEC.name
    }

    fn description(&self) -> &'static str {
        &SPEC.description
    }

    fn is_dynamic(&self) -> bool {
        SPEC.dynamic.unwrap_or(true)
    }

    async fn get(
        &self,
        runtime: &dyn IAgentRuntime,
        message: &Memory,
        _state: Option<&State>,
    ) -> PluginResult<ProviderResult> {
        let Some(room_id) = message.room_id else {
            return Ok(ProviderResult::new("")
                .with_value("messageCount", 0)
                .with_value("hasHistory", false)
                .with_data("messages", Vec::<serde_json::Value>::new()));
        };

        let mut sections = Vec::new();
        let mut message_list: Vec<serde_json::Value> = Vec::new();

        // Get recent messages
        match runtime.get_memories(Some(room_id), None, None, 20).await {
            Ok(messages) => {
                // Reverse for chronological order
                for msg in messages.into_iter().rev() {
                    if msg.content.text.is_empty() {
                        continue;
                    }

                    // Get sender name
                    let sender_name = if let Some(entity_id) = msg.entity_id {
                        if entity_id == runtime.agent_id() {
                            runtime.character().name.clone()
                        } else if let Ok(Some(entity)) = runtime.get_entity(entity_id).await {
                            entity.name.unwrap_or_else(|| "Unknown".to_string())
                        } else {
                            "Unknown".to_string()
                        }
                    } else {
                        "Unknown".to_string()
                    };

                    let mut message_text = msg.content.text.clone();
                    if message_text.len() > 300 {
                        message_text.truncate(300);
                        message_text.push_str("...");
                    }

                    message_list.push(serde_json::json!({
                        "id": msg.id.to_string(),
                        "sender": sender_name,
                        "text": message_text,
                        "timestamp": msg.created_at.timestamp()
                    }));

                    sections.push(format!("**{}**: {}", sender_name, message_text));
                }
            }
            Err(e) => {
                runtime.log_warning(
                    "provider:recentMessages",
                    &format!("Error fetching messages: {}", e),
                );
            }
        }

        let context_text = if sections.is_empty() {
            String::new()
        } else {
            format!("# Recent Messages\n{}", sections.join("\n"))
        };

        Ok(ProviderResult::new(context_text)
            .with_value("messageCount", message_list.len() as i64)
            .with_value("hasHistory", !message_list.is_empty())
            .with_value("roomId", room_id.to_string())
            .with_data("messages", message_list)
            .with_data("roomId", room_id.to_string()))
    }
}
